godot-core/src/builtin/meta/godot_compat/mod.rs

```diff
@@ -0,0 +1,120 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+mod impls;
+
+use godot_ffi as sys;
+
+use crate::builtin::{Variant, VariantConversionError};
+
+use super::{GodotFfiVariant, GodotType};
+
+/// Indicates that a type has some canonical Godot type that can represent it.
+///
+/// The type specified here is what will be used to pass this type across to ffi-boundary to/from Godot.
+/// Generally [`ToGodot`] needs to be implemented to pass a type to Godot, and [`FromGodot`] to receive this
+/// type from Godot.
+pub trait GodotCompatible {
+    /// The type used for ffi-passing.
+    type Via: GodotType;
+}
+
+/// Defines the canonical conversion to Godot for a type.
+///
+/// It is assumed that all the methods return equal values given equal inputs. Additionally it is assumed
+/// that if [`FromGodot`] is implemented, converting to Godot and back again will return a value equal to the
+/// starting value.
+///
+/// Violating these assumptions is safe but will give unexpected results.
+pub trait ToGodot: Sized + GodotCompatible {
+    /// Converts this type to the Godot type by reference, usually by cloning.
+    fn to_godot(&self) -> Self::Via;
+
+    /// Converts this type to the Godot type.
+    fn into_godot(self) -> Self::Via {
+        self.to_godot()
+    }
+
+    /// Converts this type to a [Variant].
+    fn to_variant(&self) -> Variant {
+        self.to_godot().to_ffi().ffi_to_variant()
+    }
+}
+
+/// Defines the canonical conversion from Godot for a type.
+///
+/// It is assumed that all the methods return equal values given equal inputs. Additionally it is assumed
+/// that if [`ToGodot`] is implemented, converting to Godot and back again will return a value equal to the
+/// starting value.
+///
+/// Violating these assumptions is safe but will give unexpected results.
+pub trait FromGodot: Sized + GodotCompatible {
+    // TODO: better error
+    /// Performs the conversion.
+    fn try_from_godot(via: Self::Via) -> Option<Self>;
```

Comment:
Even if we don't have an error yet, should we not return `Result<Self, ()>` for now? That would just mean changing the `Err` type later, rather than changing from `Option` to `Result`.

It would also come with `#[must_use]` etc.

Reply:
Makes sense yeah.

Reply:
Actually thinking about it, it might be better to not do that. I can add a `#[must_use]`, but turning this into a `Result<Self, ()>` means that i have to change every use-case to handling `Result`s instead usually by ignoring the `Err` value. Which means that when we change this later we can't rely on the compiler to tell us where we're not properly handling the error value, since we're just ignoring them everywhere.

Reply:
The reason this PR doesn't contain the change to using `Result` initially is that it requires us to change a lot of code to make sure the `Err` is propagated properly and all error info is preserved throughout call chains where desired. 

So instead of doing that, this PR is currently just failing with generic error messages in those cases. If we then change to `Result` in preparation for the future change, then we'd either need to just ignore the `Err` and fail with generic error messages still, or we'd need to actually make sure `Err` values are propagated through everything as they should. But propagating `Err` as we should was what i was trying to avoid doing so that the PR wouldn't get too big in the first place. And just ignoring `Err` and failing with generic error messages isn't really much better than just using `Option`. And it also means that we're more likely to miss result handling when we do make the change in the future.

I was planning on working on the error handling after this PR btw so it should hopefully not be too long a delay either way.

Reply:
I see. Makes sense to split it then to a later change, thanks for elaboration!

Either way, it's not very urgent. Having this PR is a great base! 😊